[WebXR][OpenXR] Add input support

svillar opened the pull request:
#### befb287e6ce7defbdcb7253a94f51a8a262a59d9
<pre>
[WebXR][OpenXR] Add input support
<a href="https://bugs.webkit.org/show_bug.cgi?id=297951">https://bugs.webkit.org/show_bug.cgi?id=297951</a>

Reviewed by Carlos Garcia Campos.

This implements the OpenXR machinery required to retrieve the
available input methods for WebXR sessions as long as the
information associated to them, like poses, status of buttons,
axes, etc...

The OpenXR coordinator just need to instance an OpenXRInput object
and then ask it to collect the input sources information on every
frame. The OpenXRInput object in turn instances as many OpenXRInputSources
as required, typically two (one for each hand). Those input sources
are the ones that finally query the state of the defined input actions.

In order to have all the interaction profiles information in a single place
a OpenXRInputMappings header file was created. That would store all the
information about the supported interaction profiles and their suggested
bindings for the different hardware components they support. So far we
define two supported profiles, the hand interaction profile (useful for
hand tracking only devices) and the generic simple profile from Khronos
which must be supported by all OpenXR runtimes.

These changes required converting the OpenXRExtensions object into a
singleton as it needs to be accessed by the OpenXRInputSources as well
(to check the availability of extensions).

* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/XR/openxr/OpenXRExtensions.cpp:
(WebKit::OpenXRExtensions::singleton):
(WebKit::OpenXRExtensions::OpenXRExtensions):
(WebKit::OpenXRExtensions::create): Deleted.
* Source/WebKit/UIProcess/XR/openxr/OpenXRExtensions.h:
* Source/WebKit/UIProcess/XR/openxr/OpenXRInput.cpp: Added.
(WebKit::OpenXRInput::create):
(WebKit::OpenXRInput::OpenXRInput):
(WebKit::OpenXRInput::initialize):
(WebKit::OpenXRInput::collectInputSources const):
(WebKit::OpenXRInput::updateInteractionProfile):
* Source/WebKit/UIProcess/XR/openxr/OpenXRInput.h: Added.
* Source/WebKit/UIProcess/XR/openxr/OpenXRInputMappings.h: Added.
(WebKit::buttonTypeToString):
(WebKit::axisTypetoString):
* Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp: Added.
(WebKit::OpenXRInputSource::create):
(WebKit::OpenXRInputSource::OpenXRInputSource):
(WebKit::OpenXRInputSource::~OpenXRInputSource):
(WebKit::OpenXRInputSource::initialize):
(WebKit::OpenXRInputSource::suggestBindings const):
(WebKit::OpenXRInputSource::collectInputSource const):
(WebKit::OpenXRInputSource::updateInteractionProfile):
(WebKit::OpenXRInputSource::createActionSpace const):
(WebKit::OpenXRInputSource::createAction const):
(WebKit::OpenXRInputSource::createButtonActions const):
(WebKit::OpenXRInputSource::createBinding const):
(WebKit::OpenXRInputSource::getPose const):
(WebKit::OpenXRInputSource::collectButton const):
(WebKit::OpenXRInputSource::collectAxis const):
(WebKit::OpenXRInputSource::getActionState const):
* Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.h: Added.
(WebKit::OpenXRInputSource::actionSet const):
* Source/WebKit/UIProcess/XR/openxr/OpenXRUtils.h:
(WebKit::handednessToString):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::getPrimaryDeviceInfo):
(WebKit::OpenXRCoordinator::startSession):
(WebKit::OpenXRCoordinator::createInstance):
(WebKit::OpenXRCoordinator::initializeDevice):
(WebKit::OpenXRCoordinator::tryInitializeGraphicsBinding):
(WebKit::OpenXRCoordinator::createSessionIfNeeded):
(WebKit::OpenXRCoordinator::cleanupSessionAndAssociatedResources):
(WebKit::OpenXRCoordinator::handleSessionStateChange):
(WebKit::OpenXRCoordinator::pollEvents):
(WebKit::OpenXRCoordinator::populateFrameData):
(WebKit::OpenXRCoordinator::createReferenceSpacesIfNeeded):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:

Canonical link: <a href="https://commits.webkit.org/299200@main">https://commits.webkit.org/299200@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6291e8c63281f1787b7039308e91f05997858f2e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118237 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37917 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28559 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124399 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70284 "Built successfully") 
| | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38612 "Hash 6291e8c6 for PR 49924 does not build (failure)") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46499 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/89726 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121190 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/130/builds/38612 "Hash 6291e8c6 for PR 49924 does not build (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106014 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70219 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/130/builds/38612 "Hash 6291e8c6 for PR 49924 does not build (failure)") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68063 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/130/builds/38612 "Hash 6291e8c6 for PR 49924 does not build (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24311 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127472 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45143 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34034 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/98408 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45505 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102233 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98194 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24962 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43589 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21581 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/41613 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45013 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50689 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44475 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47820 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46163 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInput.cpp:
```diff
@@ -0,0 +1,106 @@
+XrResult OpenXRInput::initialize()
+{
+    for (auto handedness : { PlatformXR::XRHandedness::Left, PlatformXR::XRHandedness::Right }) {
+        m_handleIndex++;;
```
```suggestion
        m_handleIndex++;
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInput.cpp:
```diff
@@ -0,0 +1,106 @@
+
+std::unique_ptr<OpenXRInput> OpenXRInput::create(XrInstance instance, XrSession session)
+{
+    auto input = std::unique_ptr<OpenXRInput>(new OpenXRInput(instance, session));
```
Could we use makeUnique?

svillar replied:
No because the constructor is private

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInput.cpp:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    OpenXRInputSource::SuggestedBindings bindings;
+    Vector<XrActionSet> actionSets;
```
We can se initialize capacity to m_inputSources.size()

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInput.cpp:
```diff
@@ -0,0 +1,106 @@
+
+    OpenXRInputSource::SuggestedBindings bindings;
+    Vector<XrActionSet> actionSets;
+    for (auto& inputSource : m_inputSources) {
```
```suggestion
    for (const auto& inputSource : m_inputSources) {
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInput.cpp:
```diff
@@ -0,0 +1,106 @@
+        actionSets.append(inputSource->actionSet());
+    }
+
+    for (auto& binding : bindings) {
```
```suggestion
    for (const auto& binding : bindings) {
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInput.cpp:
```diff
@@ -0,0 +1,106 @@
+
+Vector<PlatformXR::FrameData::InputSource> OpenXRInput::collectInputSources(const XrFrameState& frameState, XrSpace space) const
+{
+    Vector<XrActiveActionSet> actionSets;
```
We can set initial capacity here too.

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInput.cpp:
```diff
@@ -0,0 +1,106 @@
+Vector<PlatformXR::FrameData::InputSource> OpenXRInput::collectInputSources(const XrFrameState& frameState, XrSpace space) const
+{
+    Vector<XrActiveActionSet> actionSets;
+    for (auto& input : m_inputSources)
```
```suggestion
    for (const auto& input : m_inputSources)
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInput.cpp:
```diff
@@ -0,0 +1,106 @@
+    syncInfo.activeActionSets = actionSets.span().data();
+    CHECK_XRCMD(xrSyncActions(m_session, &syncInfo));
+
+    Vector<PlatformXR::FrameData::InputSource> result;
```
Same here.

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInput.cpp:
```diff
@@ -0,0 +1,106 @@
+    CHECK_XRCMD(xrSyncActions(m_session, &syncInfo));
+
+    Vector<PlatformXR::FrameData::InputSource> result;
+    for (auto& input : m_inputSources) {
```
```suggestion
    for (const auto& input : m_inputSources) {
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputMappings.h:
```diff
@@ -0,0 +1,154 @@
+    OpenXRButtonType::ButtonA, OpenXRButtonType::ButtonB
+};
+
+constexpr ASCIILiteral kPathSelect { "/input/select"_s };
```
We don't normally use kFoo in WebKit, we normally use s_foo or just foo.

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputMappings.h:
```diff
@@ -0,0 +1,154 @@
+    }
+}
+
+enum OpenXRButtonFlags {
```
Can this be enum class?

svillar replied:
We could but we'd have to add some extra machinery to perform flag checks. As it is now we can easily just use the and operator

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputMappings.h:
```diff
@@ -0,0 +1,154 @@
+enum OpenXRButtonFlags {
+    Click = (1u << 0),
+    Touch = (1u << 1),
+    Value  = (1u << 2),
```
```suggestion
    Click = 1u << 0,
    Touch = 1u << 1,
    Value  = 1u << 2,
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputMappings.h:
```diff
@@ -0,0 +1,154 @@
+    Value  = (1u << 2),
+};
+
+enum OpenXRHandFlags {
```
Can this be enum class?

svillar replied:
Sure

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputMappings.h:
```diff
@@ -0,0 +1,154 @@
+
+enum OpenXRHandFlags {
+    Left = (1u << 0),
+    Right = (1u << 1),
```
```suggestion
    Left = 1u << 0,
    Right = 1u << 1,
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputMappings.h:
```diff
@@ -0,0 +1,154 @@
+};
+
+struct OpenXRButton {
+    OpenXRButtonType type;
+    OpenXRButtonPath path;
+    OpenXRButtonFlags flags;
+    OpenXRHandFlags hand;
```
Can we initialize all values?

svillar replied:
Not really, there are nothing such a default value for them, they heavily depend on the hardware that is being used.

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputMappings.h:
```diff
@@ -0,0 +1,154 @@
+}
+
+struct OpenXRAxis {
+    OpenXRAxisType type;
+    OpenXRButtonPath path;
```
Can we initialize these too?

svillar replied:
Same

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputMappings.h:
```diff
@@ -0,0 +1,154 @@
+    std::span<const OpenXRAxis> axes;
+};
+
+constexpr std::array<OpenXRProfileId, 3> HandInteractionProfileIds { "generic-hand-select-grasp", "generic-hand-select", "generic-hand" };
```
```suggestion
constexpr std::array<OpenXRProfileId, 3> handInteractionProfileIds { "generic-hand-select-grasp", "generic-hand-select", "generic-hand" };
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputMappings.h:
```diff
@@ -0,0 +1,154 @@
+};
+
+constexpr std::array<OpenXRProfileId, 3> HandInteractionProfileIds { "generic-hand-select-grasp", "generic-hand-select", "generic-hand" };
+constexpr std::array<OpenXRButton, 2> HandInteractionProfileButtons {
```
```suggestion
constexpr std::array<OpenXRButton, 2> handInteractionProfileButtons {
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputMappings.h:
```diff
@@ -0,0 +1,154 @@
+    OpenXRButton { .type = OpenXRButtonType::Squeeze, .path = kPathGraspExt, .flags = OpenXRButtonFlags::Value, .hand = OpenXRHandFlags::Both },
+};
+
+constexpr OpenXRInteractionProfile HandInteractionProfile {
```
```suggestion
constexpr OpenXRInteractionProfile handInteractionProfile {
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputMappings.h:
```diff
@@ -0,0 +1,154 @@
+};
+
+// Default fallback when there isn't a specific controller binding.
+constexpr std::array<ASCIILiteral, 1> KHRSimpleProfileIds { "generic-button"_s };
```
```suggestion
constexpr std::array<ASCIILiteral, 1> khrSimpleProfileIds { "generic-button"_s };
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputMappings.h:
```diff
@@ -0,0 +1,154 @@
+// Default fallback when there isn't a specific controller binding.
+constexpr std::array<ASCIILiteral, 1> KHRSimpleProfileIds { "generic-button"_s };
+
+constexpr std::array<OpenXRButton, 1> KHRSimpleButtons {
```
```suggestion
constexpr std::array<OpenXRButton, 1> khrSimpleButtons {
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputMappings.h:
```diff
@@ -0,0 +1,154 @@
+    OpenXRButton { .type = OpenXRButtonType::Trigger, .path = kPathSelect, .flags = OpenXRButtonFlags::Click, .hand = OpenXRHandFlags::Both }
+};
+
+constexpr OpenXRInteractionProfile KHRSimpleControllerProfile {
```
```suggestion
constexpr OpenXRInteractionProfile khrSimpleControllerProfile {
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp:
```diff
@@ -0,0 +1,400 @@
+/*
+ * Copyright (C) 2021 Igalia, S.L.
```
```suggestion
 * Copyright (C) 2025 Igalia, S.L.
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp:
```diff
@@ -0,0 +1,400 @@
+constexpr auto kInputGripPath { "/input/grip/pose"_s };
+constexpr auto kInputAimPath { "/input/aim/pose"_s };
+constexpr auto kInputPinchPath { "/input/pinch_ext/pose"_s };
+constexpr auto kInputPokePath { "/input/poke_ext/pose"_s };
```
Same here about kFoo

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp:
```diff
@@ -0,0 +1,400 @@
+
+std::unique_ptr<OpenXRInputSource> OpenXRInputSource::create(XrInstance instance, XrSession session, PlatformXR::XRHandedness handedness, PlatformXR::InputSourceHandle handle)
+{
+    auto input = std::unique_ptr<OpenXRInputSource>(new OpenXRInputSource(instance, session, handedness, handle));
```
Can we use makeUnique?

svillar replied:
No because the constructor is private

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp:
```diff
@@ -0,0 +1,403 @@
+        }
+    };
+
+    for (auto& profile : openXRInteractionProfiles) {
```
```suggestion
    for (const auto& profile : openXRInteractionProfiles) {
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp:
```diff
@@ -0,0 +1,403 @@
+    return XR_SUCCESS;
+}
+
+std::optional<PlatformXR::FrameData::InputSource> OpenXRInputSource::getInputSource(XrSpace localSpace, const XrFrameState& frameState) const
```
```suggestion
std::optional<PlatformXR::FrameData::InputSource> OpenXRInputSource::inputSource(XrSpace localSpace, const XrFrameState& frameState) const
```

svillar replied:
Finally replaced with collectInputSource. It's coherent with the method in OpenXRInput and also it does not give the impression that is a simple getter.

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp:
```diff
@@ -0,0 +1,403 @@
+    if (XR_SUCCEEDED(getPose(m_gripSpace, localSpace, frameState, gripPose)))
+        data.gripOrigin = gripPose;
+
+    Vector<std::optional<PlatformXR::FrameData::InputSourceButton>> buttons;
```
We can set initial capacity here too, but I think the size is known at compile time, so we could even set inline capacity instead.

svillar replied:
It's unlikely to have all the button types actually available, but I guess it still pays off knowing that the amount of them is very low

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp:
```diff
@@ -0,0 +1,403 @@
+        }
+    }
+
+    Vector<std::optional<XrVector2f>> axes;
```
Same here

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp:
```diff
@@ -0,0 +1,403 @@
+    return XR_SUCCESS;
+}
+
+std::optional<PlatformXR::FrameData::InputSourceButton> OpenXRInputSource::getButton(OpenXRButtonType buttonType) const
```
```suggestion
std::optional<PlatformXR::FrameData::InputSourceButton> OpenXRInputSource::button(OpenXRButtonType buttonType) const
```

svillar replied:
Using also collect

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp:
```diff
@@ -0,0 +1,403 @@
+    return hasValue ?  std::make_optional(result) : std::nullopt;
+}
+
+std::optional<XrVector2f> OpenXRInputSource::getAxis(OpenXRAxisType axisType) const
```
```suggestion
std::optional<XrVector2f> OpenXRInputSource::axis(OpenXRAxisType axisType) const
```

svillar replied:
Ditto collect

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.h:
```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (C) 2021 Igalia, S.L.
```
```suggestion
 * Copyright (C) 2025 Igalia, S.L.
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.h:
```diff
@@ -0,0 +1,92 @@
+
+#include "OpenXRInputMappings.h"
+#include "OpenXRUtils.h"
+
```
```suggestion
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRUtils.h:
```diff
@@ -86,6 +95,20 @@ inline PlatformXR::FrameData::View XrViewToView(XrView view)
     return { XrPosefToPose(view.pose), PlatformXR::FrameData::Fov { std::abs(view.fov.angleUp), std::abs(view.fov.angleDown), std::abs(view.fov.angleLeft), std::abs(view.fov.angleRight) } };
 }
 
+inline String handednessToString(PlatformXR::XRHandedness handedness)
```
Could this return ASCIILiteral?

svillar replied:
I guess so

spenap commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputMappings.h:
```diff
@@ -0,0 +1,154 @@
+    std::span<const OpenXRAxis> axes;
+};
+
+constexpr std::array<OpenXRProfileId, 3> handInteractionProfileIds { "generic-hand-select-grasp", "generic-hand-select", "generic-hand" };
```
```suggestion
constexpr std::array<OpenXRProfileId, 3> handInteractionProfileIds { "generic-hand-select-grasp"_s, "generic-hand-select"_s, "generic-hand"_s };
```

or is it there an exception in this case?

svillar replied:
not needed apparently

spenap commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp:
```diff
@@ -0,0 +1,403 @@
+        data.gripOrigin = gripPose;
+
+    Vector<std::optional<PlatformXR::FrameData::InputSourceButton>, openXRButtonTypes.size()> buttons;
+    for (auto& type : openXRButtonTypes) {
```
```suggestion
    for (auto type : openXRButtonTypes) {
```

For consistency, it's iterated by value in the other occasion, and it's not being modified

spenap commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp:
```diff
@@ -0,0 +1,403 @@
+            data.buttons.append(*buttons[i]);
+            continue;
+        }
+        // Add placeholder if there are more valid buttons in the list.
```
How does this work? If a valid button is found, we continue, so we don't reach here.

If we find a button without value, then we try to find the next valid one, and we append an empty placeholder, then we break from that loop... but go back to the outer loop. 

Then in the outer loop we would eventually find the same valid buttons, but we would have gaps? 
buttons:
[V] [I] [I] [V]

data.buttons:
[V] [ ] [ ] [V]
?

svillar replied:
Yes because both the OpenXR and WebXR specs define a precise order for the different buttons. If you don't have them you just leave gaps

spenap replied:
I see.

I think you could express the same with

```cpp

data.buttons = Vector<InputSourceButton>(buttons.size(), {});
for (size_t i = 0; i < buttons.size(); ++i) {
    if (buttons[i])
        data.buttons[i] = buttons[i];
}
```

Or maybe this works?...

```cpp
for (size_t i = 0; i < buttons.size(); ++i) {
    data.buttons[i] = buttons[i].has_value() ? buttons[i] : {};
}
```

spenap commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp:
```diff
@@ -0,0 +1,403 @@
+    }
+
+    Vector<std::optional<XrVector2f>, openXRAxisTypes.size()> axes;
+    for (auto& type : openXRAxisTypes)
```
here I would again make it by value

```suggestion
    for (auto type : openXRAxisTypes)
```

as there's no benefit in const-ref, and we don't seem to need ref

spenap commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp:
```diff
@@ -0,0 +1,403 @@
+            data.axes.append(axes[i]->y);
+            continue;
+        }
+        // Add placeholder if there are more valid axes in the list.
```
I guess this is the same approach as with the buttons, right?

svillar replied:
yep

spenap commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp:
```diff
@@ -0,0 +1,403 @@
+    RETURN_RESULT_IF_FAILED(xrPathToString(m_instance, state.interactionProfile, bufferSize, &writtenCount, buffer));
+
+    m_profiles.clear();
+    for (auto& profile : openXRInteractionProfiles) {
```
do we need to take `profile` by reference?

svillar replied:
just to avoid copies

spenap replied:
this is a lightweight object, an enum. I personally go for value in these cases, but no strong opinion. What I would say is to go for const ref if you want to keep a ref

spenap commented on Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp:
```diff
@@ -0,0 +1,403 @@
+
+    auto state = createOpenXRStruct<XrActionStateBoolean, XR_TYPE_ACTION_STATE_BOOLEAN>();
+    auto info = createOpenXRStruct<XrActionStateGetInfo, XR_TYPE_ACTION_STATE_GET_INFO>();
+    info.action = action;
+
+    RETURN_RESULT_IF_FAILED(xrGetActionStateBoolean(m_session, &info, &state));
+    *value = state.currentState;
+
+    return XR_SUCCESS;
+}
+
+XrResult OpenXRInputSource::getActionState(XrAction action, float* value) const
+{
+    ASSERT(value);
+    ASSERT(action != XR_NULL_HANDLE);
+
+    auto state = createOpenXRStruct<XrActionStateFloat, XR_TYPE_ACTION_STATE_FLOAT>();
+    auto info = createOpenXRStruct<XrActionStateGetInfo, XR_TYPE_ACTION_STATE_GET_INFO>();
+    info.action = action;
+
+    RETURN_RESULT_IF_FAILED(xrGetActionStateFloat(m_session, &info, &state));
+    *value = state.currentState;
+
+    return XR_SUCCESS;
+}
+
+XrResult OpenXRInputSource::getActionState(XrAction action, XrVector2f* value) const
+{
+    ASSERT(value);
+    ASSERT(action != XR_NULL_HANDLE);
+
+    auto state = createOpenXRStruct<XrActionStateVector2f, XR_TYPE_ACTION_STATE_VECTOR2F>();
+    auto info = createOpenXRStruct<XrActionStateGetInfo, XR_TYPE_ACTION_STATE_GET_INFO>();
+    info.action = action;
+
+    RETURN_RESULT_IF_FAILED(xrGetActionStateVector2f(m_session, &info, &state));
+    *value = state.currentState;
+
+    return XR_SUCCESS;
+}
```
totally not for this PR, but these 3 could be a template method, I think

svillar replied:
good suggestion

svillar committed:
[WebXR][OpenXR] Add input support
https://bugs.webkit.org/show_bug.cgi?id=297951

Reviewed by Carlos Garcia Campos.

This implements the OpenXR machinery required to retrieve the
available input methods for WebXR sessions as long as the
information associated to them, like poses, status of buttons,
axes, etc...

The OpenXR coordinator just need to instance an OpenXRInput object
and then ask it to collect the input sources information on every
frame. The OpenXRInput object in turn instances as many OpenXRInputSources
as required, typically two (one for each hand). Those input sources
are the ones that finally query the state of the defined input actions.

In order to have all the interaction profiles information in a single place
a OpenXRInputMappings header file was created. That would store all the
information about the supported interaction profiles and their suggested
bindings for the different hardware components they support. So far we
define two supported profiles, the hand interaction profile (useful for
hand tracking only devices) and the generic simple profile from Khronos
which must be supported by all OpenXR runtimes.

These changes required converting the OpenXRExtensions object into a
singleton as it needs to be accessed by the OpenXRInputSources as well
(to check the availability of extensions).

* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/XR/openxr/OpenXRExtensions.cpp:
(WebKit::OpenXRExtensions::singleton):
(WebKit::OpenXRExtensions::OpenXRExtensions):
(WebKit::OpenXRExtensions::create): Deleted.
* Source/WebKit/UIProcess/XR/openxr/OpenXRExtensions.h:
* Source/WebKit/UIProcess/XR/openxr/OpenXRInput.cpp: Added.
(WebKit::OpenXRInput::create):
(WebKit::OpenXRInput::OpenXRInput):
(WebKit::OpenXRInput::initialize):
(WebKit::OpenXRInput::collectInputSources const):
(WebKit::OpenXRInput::updateInteractionProfile):
* Source/WebKit/UIProcess/XR/openxr/OpenXRInput.h: Added.
* Source/WebKit/UIProcess/XR/openxr/OpenXRInputMappings.h: Added.
(WebKit::buttonTypeToString):
(WebKit::axisTypetoString):
* Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp: Added.
(WebKit::OpenXRInputSource::create):
(WebKit::OpenXRInputSource::OpenXRInputSource):
(WebKit::OpenXRInputSource::~OpenXRInputSource):
(WebKit::OpenXRInputSource::initialize):
(WebKit::OpenXRInputSource::suggestBindings const):
(WebKit::OpenXRInputSource::collectInputSource const):
(WebKit::OpenXRInputSource::updateInteractionProfile):
(WebKit::OpenXRInputSource::createActionSpace const):
(WebKit::OpenXRInputSource::createAction const):
(WebKit::OpenXRInputSource::createButtonActions const):
(WebKit::OpenXRInputSource::createBinding const):
(WebKit::OpenXRInputSource::getPose const):
(WebKit::OpenXRInputSource::collectButton const):
(WebKit::OpenXRInputSource::collectAxis const):
(WebKit::OpenXRInputSource::getActionState const):
* Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.h: Added.
(WebKit::OpenXRInputSource::actionSet const):
* Source/WebKit/UIProcess/XR/openxr/OpenXRUtils.h:
(WebKit::handednessToString):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::getPrimaryDeviceInfo):
(WebKit::OpenXRCoordinator::startSession):
(WebKit::OpenXRCoordinator::createInstance):
(WebKit::OpenXRCoordinator::initializeDevice):
(WebKit::OpenXRCoordinator::tryInitializeGraphicsBinding):
(WebKit::OpenXRCoordinator::createSessionIfNeeded):
(WebKit::OpenXRCoordinator::cleanupSessionAndAssociatedResources):
(WebKit::OpenXRCoordinator::handleSessionStateChange):
(WebKit::OpenXRCoordinator::pollEvents):
(WebKit::OpenXRCoordinator::populateFrameData):
(WebKit::OpenXRCoordinator::createReferenceSpacesIfNeeded):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:

Canonical link: https://commits.webkit.org/299200@main
